File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        HashMap<String, Integer> person = new HashMap<>();

        // 참가자 수를 카운트
        for (String p : participant) {
            person.put(p, person.getOrDefault(p, 0) + 1);
        }

        // 완주자 수를 감소
        for (String c : completion) {
            person.put(c, person.get(c) - 1);
        }

        // 값이 1인 사람이 완주 못한 사람
        for (String key : person.keySet()) {
            if (person.get(key) > 0) {
                return key;
            }
        }

        return "";  // 이론상 도달하지 않음
    }
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.js
function solution(participant, completion) {
    // 1. 참가자 이름을 카운트할 Map 생성
    const map = new Map();

    // 2. 참가자 배열에서 이름 카운트 증가
    for (let name of participant) {
        map.set(name, (map.get(name) || 0) + 1);
    }

    // 3. 완주자 배열에서 해당 이름 카운트 감소
    for (let name of completion) {
        map.set(name, map.get(name) - 1);
    }

    // 4. 카운트가 1인 사람이 완주하지 못한 사람
    for (let [name, count] of map) {
        if (count > 0) {
            return name;
        }
    }
}
